app/helpers/mapping_helper.rb

```diff
@@ -0,0 +1,49 @@
+require 'uri'
+
+module MappingHelper
+  def user_list
+    users_from_group = ENV['USERS_FROM_GROUP']
+    result = `getent group #{users_from_group}`
+
+    # example output
+    #    "wiagstf:*:5362:mrodgers,efranz\n"
+    result.strip.split(':')[3].split(',').sort
+  end
+
+
+  # Get a list of the various Shiny apps
+  def app_list
+    glob_results = Dir.glob(
+      File.join File.expand_path(
+        ENV['APP_PROJECT_SPACE']
+      ), 'bc_shiny_*'
+    )
+
+    # Create a sorted list of just the App directory
+    glob_results.sort
+  end
+
+
+  def get_app_name app_path
+    URI(app_path).path.split('/').last
+  end
+
+
+  # Get a list of datasets
+  # The list will consist of datasets detected at the configured location and
+  # datasets that have been saved to the database but are stored in nonstandard
+  # locations.
+  def known_datasets
+    installed_datasets = Set.new Dir.glob(
+      File.join File.expand_path(
+        ENV['APP_DATASET_ROOT']
+      ), '*'
+    )
+
+    persisted_datasets = Set.new Mapping.select(:dataset).map(&:dataset)
+    non_std_location_datasets = persisted_datasets - installed_datasets
+
+    # Sort the lists, keeping the two location types separate
+    installed_datasets.to_a.sort + non_std_location_datasets.to_a.sort
+  end
```

Comment:
A functional approach to this would be to have a method `managed_datasets` or `installed_datasets` that returns the list of datasets in BISR's `dataset_root` and then do something like this:

```ruby
def known_datasets
  (installed_datasets | Mapping.select(:dataset).map(&:dataset).sort.to_a).uniq
end
```

The `|` preserves order so you get the result I think you were hoping for. Per our discussion of if there was a more appropriate place for this - you could put this particular method on the model itself in a block like:

```ruby
class Mapping

  class << self
    # class methods here
  end
```

Since the method is doing a db query better in the model than in a helper or view.

Also, while Set ensures uniqueness, in this case just using the `uniq` method on `Array` is probably good enough. I think using Set becomes beneficial if we would imagine manipulating the set after it was initialized.

Reply:
Since `installed_datasets` is sure to be unique, you could actually use Active Record's `distinct` which I think is an alias for `uniq`.

Might be more readable:

```ruby
def known_datasets
  installed_datasets | Mapping.select(:dataset).distinct.order(:dataset).map(&:dataset)
end
```

Of course in this case its best practice to use scopes (http://guides.rubyonrails.org/v4.2/active_record_querying.html#scopes) for these queries so something like:

```ruby
class Mapping
  scope :datasets, -> {  select(:dataset).distinct.order(:dataset) }
```

and then we simplify further:

```ruby
def known_datasets
  installed_datasets | Mapping.datasets.map(&:dataset)
end
```

Well maybe scope is inappropriate here, and rather just a class method to return that array of datasets, because this seems odd `Mapping.datasets.map(&:dataset)`.

```ruby
class Mapping
  def self.datasets
    select(:dataset).distinct.order(:dataset).map(&:dataset)
  end
```

then

```ruby
def known_datasets
  installed_datasets | Mapping.datasets
end
```

Anyways ideas.

Reply:
I like that `scopes` can be chained, but I find the name `scope` unintuitive. Next time I'm in a room with the ActiveRecord developers remind me to ask them why they chose that 😉.

For the dataset case I like using a static method, but instead of doing the `Enumerable.map` it looks like I should prefer `ActiveRecord.pluck`.